cranelift-codegen/src/legalizer/mod.rs

```diff
@@ -633,3 +633,34 @@ fn narrow_store(
     );
     pos.remove_inst();
 }
+
+/// Expands an illegal iconst value by splitting it into two.
+fn narrow_iconst(
+    inst: ir::Inst,
+    func: &mut ir::Function,
+    _cfg: &mut ControlFlowGraph,
+    _isa: &dyn TargetIsa,
+) {
+    let imm: i64 = if let ir::InstructionData::UnaryImm {
+        opcode: ir::Opcode::Iconst,
+        imm,
+    } = &func.dfg[inst]
+    {
+        (*imm).into()
+    } else {
+        panic!("unexpected instruction in expand_iconst");
+    };
+
+    let mut pos = FuncCursor::new(func).at_inst(inst);
+    pos.use_srcloc(inst);
+
+    if pos.func.dfg.ctrl_typevar(inst) == I64 {
+        let low = pos.ins().iconst(I32, imm & 0xffffffff);
+        let high = pos.ins().iconst(I32, imm >> 32);
+        // The instruction has as many results as iconcat, so no need to replace them.
+        pos.func.dfg.replace(inst).iconcat(low, high);
+        return;
+    }
+
+    unimplemented!();
+}
```

Comment:
Maybe I'm missing something but I don't see where `narrow_iconst` has a predicate limiting it to 32-bit architectures; on 64-bit architectures I would expect this legalization not to be run because we can effectively load 64-bit integers directly with https://github.com/CraneStation/cranelift/blob/0fd81b590af06835c6faa67a212f4c84b21028d8/cranelift-codegen/meta/src/isa/x86/encodings.rs#L729.

Reply:
Good point; this doesn't run on x86 64-bits, but it shouldn't probably either run for future 64-bits architectures like aarch64. Will add an ISA check in this function.